--- binutils-2.25.1.new/ld/testsuite/ld-powerpc/tlsso32.r	2016-03-01 14:08:17.249137110 +0000
+++ binutils-2.25.1/ld/testsuite/ld-powerpc/tlsso32.r	2016-03-01 15:10:02.847812018 +0000
@@ -20,6 +20,7 @@ Section Headers:
  +\[[ 0-9]+\] \.dynamic +DYNAMIC .* 08 +WA +3 +0 +4
  +\[[ 0-9]+\] \.got +PROGBITS .* 0+30 04 +WA +0 +0 +4
  +\[[ 0-9]+\] \.plt +PROGBITS .* 0+4 00 +WA +0 +0 +4
+#...
  +\[[ 0-9]+\] \.shstrtab +.*
  +\[[ 0-9]+\] \.symtab +.*
  +\[[ 0-9]+\] \.strtab +.*
@@ -80,12 +81,12 @@ Symbol table '\.dynsym' contains [0-9]+
 .* TLS +GLOBAL +DEFAULT +8 le1
 .* TLS +GLOBAL +DEFAULT +UND ld
 .* NOTYPE +GLOBAL +DEFAULT +6 _start
-.* NOTYPE +GLOBAL +DEFAULT +11 __end
+.* NOTYPE +GLOBAL +DEFAULT +.. __end
 .* TLS +GLOBAL +DEFAULT +8 ld2
 .* TLS +GLOBAL +DEFAULT +8 ld1
-.* NOTYPE +GLOBAL +DEFAULT +11 __bss_start
-.* NOTYPE +GLOBAL +DEFAULT +11 _edata
-.* NOTYPE +GLOBAL +DEFAULT +11 _end
+.* NOTYPE +GLOBAL +DEFAULT +.. __bss_start
+.* NOTYPE +GLOBAL +DEFAULT +.. _edata
+.* NOTYPE +GLOBAL +DEFAULT +.. _end
 .* TLS +GLOBAL +DEFAULT +8 gd0
 .* TLS +GLOBAL +DEFAULT +8 ie0
 
@@ -103,32 +104,4 @@ Symbol table '\.symtab' contains [0-9]+
 .* SECTION +LOCAL +DEFAULT +9 
 .* SECTION +LOCAL +DEFAULT +10 
 .* SECTION +LOCAL +DEFAULT +11 
-.* FILE +LOCAL +DEFAULT +ABS .*
-.* TLS +LOCAL +DEFAULT +7 gd4
-.* TLS +LOCAL +DEFAULT +7 ld4
-.* TLS +LOCAL +DEFAULT +7 ld5
-.* TLS +LOCAL +DEFAULT +7 ld6
-.* TLS +LOCAL +DEFAULT +7 ie4
-.* TLS +LOCAL +DEFAULT +7 le4
-.* TLS +LOCAL +DEFAULT +7 le5
-.* FILE +LOCAL +DEFAULT +ABS .*
-.* OBJECT +LOCAL +DEFAULT +9 _DYNAMIC
-.* NOTYPE +LOCAL +DEFAULT +6 0+8000\.got2\.plt_pic32\.__tls_get_addr
-.* NOTYPE +LOCAL +DEFAULT +6 __glink_PLTresolve
-.* NOTYPE +LOCAL +DEFAULT +6 __glink
-.* OBJECT +LOCAL +DEFAULT +10 _GLOBAL_OFFSET_TABLE_
-.* TLS +GLOBAL +DEFAULT +UND gd
-.* TLS +GLOBAL +DEFAULT +8 le0
-.* NOTYPE +GLOBAL +DEFAULT +UND __tls_get_addr
-.* TLS +GLOBAL +DEFAULT +8 ld0
-.* TLS +GLOBAL +DEFAULT +8 le1
-.* TLS +GLOBAL +DEFAULT +UND ld
-.* NOTYPE +GLOBAL +DEFAULT +6 _start
-.* NOTYPE +GLOBAL +DEFAULT +11 __end
-.* TLS +GLOBAL +DEFAULT +8 ld2
-.* TLS +GLOBAL +DEFAULT +8 ld1
-.* NOTYPE +GLOBAL +DEFAULT +11 __bss_start
-.* NOTYPE +GLOBAL +DEFAULT +11 _edata
-.* NOTYPE +GLOBAL +DEFAULT +11 _end
-.* TLS +GLOBAL +DEFAULT +8 gd0
-.* TLS +GLOBAL +DEFAULT +8 ie0
+#pass
--- binutils-2.25.1.new/ld/testsuite/ld-powerpc/tlsexe32.r	2016-03-01 14:08:17.248137105 +0000
+++ binutils-2.25.1/ld/testsuite/ld-powerpc/tlsexe32.r	2016-03-01 15:08:57.693493230 +0000
@@ -22,6 +22,7 @@ Section Headers:
  +\[[ 0-9]+\] \.dynamic +DYNAMIC +[0-9a-f]+ [0-9a-f]+ [0-9a-f]+ 08 +WA +4 +0 +4
  +\[[ 0-9]+\] \.got +PROGBITS +[0-9a-f]+ [0-9a-f]+ 000018 04 +WA +0 +0 +4
  +\[[ 0-9]+\] \.plt +PROGBITS +[0-9a-f]+ [0-9a-f]+ 000004 00 +WA +0 +0 +4
+#...
  +\[[ 0-9]+\] \.shstrtab +STRTAB +.*
  +\[[ 0-9]+\] \.symtab +SYMTAB +.*
  +\[[ 0-9]+\] \.strtab +STRTAB +.*
@@ -64,11 +65,11 @@ Symbol table '\.dynsym' contains [0-9]+
 .* NOTYPE +LOCAL +DEFAULT +UND 
 .* TLS +GLOBAL +DEFAULT +UND gd
 .* TLS +GLOBAL +DEFAULT +UND ld
-.* NOTYPE +GLOBAL +DEFAULT +12 __end
-.* NOTYPE +GLOBAL +DEFAULT +12 __bss_start
+.* NOTYPE +GLOBAL +DEFAULT +.. __end
+.* NOTYPE +GLOBAL +DEFAULT +.. __bss_start
 .* FUNC +GLOBAL +DEFAULT +UND __tls_get_addr_opt
-.* NOTYPE +GLOBAL +DEFAULT +12 _edata
-.* NOTYPE +GLOBAL +DEFAULT +12 _end
+.* NOTYPE +GLOBAL +DEFAULT +.. _edata
+.* NOTYPE +GLOBAL +DEFAULT +.. _end
 
 Symbol table '\.symtab' contains [0-9]+ entries:
  +Num: +Value +Size +Type +Bind +Vis +Ndx +Name
@@ -85,29 +86,5 @@ Symbol table '\.symtab' contains [0-9]+
 .* SECTION +LOCAL +DEFAULT +10 
 .* SECTION +LOCAL +DEFAULT +11 
 .* SECTION +LOCAL +DEFAULT +12 
-.* FILE +LOCAL +DEFAULT +ABS .*
-.* TLS +LOCAL +DEFAULT +8 gd4
-.* TLS +LOCAL +DEFAULT +8 ld4
-.* TLS +LOCAL +DEFAULT +8 ld5
-.* TLS +LOCAL +DEFAULT +8 ld6
-.* TLS +LOCAL +DEFAULT +8 ie4
-.* TLS +LOCAL +DEFAULT +8 le4
-.* TLS +LOCAL +DEFAULT +8 le5
-.* FILE +LOCAL +DEFAULT +ABS .*
-.* OBJECT +LOCAL +DEFAULT +10 _DYNAMIC
-.* OBJECT +LOCAL +DEFAULT +11 _GLOBAL_OFFSET_TABLE_
-.* TLS +GLOBAL +DEFAULT +UND gd
-.* TLS +GLOBAL +DEFAULT +9 le0
-.* TLS +GLOBAL +DEFAULT +9 ld0
-.* TLS +GLOBAL +DEFAULT +9 le1
-.* TLS +GLOBAL +DEFAULT +UND ld
-.* NOTYPE +GLOBAL +DEFAULT +7 _start
-.* NOTYPE +GLOBAL +DEFAULT +12 __end
-.* TLS +GLOBAL +DEFAULT +9 ld2
-.* TLS +GLOBAL +DEFAULT +9 ld1
-.* NOTYPE +GLOBAL +DEFAULT +12 __bss_start
-.* FUNC +GLOBAL +DEFAULT +UND __tls_get_addr_opt
-.* NOTYPE +GLOBAL +DEFAULT +12 _edata
-.* NOTYPE +GLOBAL +DEFAULT +12 _end
-.* TLS +GLOBAL +DEFAULT +9 gd0
-.* TLS +GLOBAL +DEFAULT +9 ie0
+#pass
+
diff -rup binutils-2.27.orig/ld/testsuite/ld-powerpc/powerpc.exp binutils-2.27/ld/testsuite/ld-powerpc/powerpc.exp
--- binutils-2.27.orig/ld/testsuite/ld-powerpc/powerpc.exp	2017-01-16 17:27:46.697628285 +0000
+++ binutils-2.27/ld/testsuite/ld-powerpc/powerpc.exp	2017-01-16 17:34:34.684104074 +0000
@@ -102,13 +102,13 @@ set ppcelftests {
      {{objdump -hw reloc.d}} "reloc.so"}
     {"APUinfo section processing" "-melf32ppc" ""
      "-a32 -me500" {apuinfo1.s apuinfo-nul.s apuinfo2.s}
-     {{readelf -x2 apuinfo.rd}} "apuinfo"}
+     {{readelf -x.PPC.EMB.apuinfo apuinfo.rd}} "apuinfo"}
     {"APUinfo VLE section processing" "-melf32ppc" ""
      "-a32 -me500 -mvle" {apuinfo1.s apuinfo-vle.s apuinfo-vle2.s}
-     {{readelf -x2 apuinfo-vle.rd}} "apuinfo-vle"}
+     {{readelf -x.PPC.EMB.apuinfo apuinfo-vle.rd}} "apuinfo-vle"}
     {"APUinfo NULL section processing" "-melf32ppc" ""
      "-a32 -me500" {apuinfo-nul1.s apuinfo-nul.s}
-     {{readelf -x2 apuinfo-nul.rd}} "apuinfo"}
+     {{readelf -x.PPC.EMB.apuinfo apuinfo-nul.rd}} "apuinfo"}
     {"TLS32 static exec" "-melf32ppc" "" "-a32"  {tls32.s tlslib32.s}
      {{objdump -dr tls32.d} {objdump -sj.got tls32.g}
       {objdump -sj.tdata tls32.t}}
