--- ld/emultempl/armelf.em~	2015-03-18 17:32:36.436208938 -0400
+++ ld/emultempl/armelf.em	2015-03-18 17:32:54.455591126 -0400
@@ -53,6 +53,7 @@
   input_flags.dynamic = ${DYNAMIC_LINK-TRUE};
   config.has_shared = `if test -n "$GENERATE_SHLIB_SCRIPT" ; then echo TRUE ; else echo FALSE ; fi`;
   config.separate_code = `if test "x${SEPARATE_CODE}" = xyes ; then echo TRUE ; else echo FALSE ; fi`;
+  link_info.relro = TRUE;
 }
 
 static void
--- ld/emultempl/aarch64elf.em~	2015-03-18 17:32:08.347172011 -0400
+++ ld/emultempl/aarch64elf.em	2015-03-18 17:32:30.866399906 -0400
@@ -40,6 +40,7 @@
   input_flags.dynamic = ${DYNAMIC_LINK-TRUE};
   config.has_shared = `if test -n "$GENERATE_SHLIB_SCRIPT" ; then echo TRUE ; else echo FALSE ; fi`;
   config.separate_code = `if test "x${SEPARATE_CODE}" = xyes ; then echo TRUE ; else echo FALSE ; fi`;
+  link_info.relro = TRUE;
 }
 
 static void
