diff -rup binutils.orig/ld/testsuite/ld-elf/shared.exp binutils-2.27/ld/testsuite/ld-elf/shared.exp
--- binutils.orig/ld/testsuite/ld-elf/shared.exp	2017-02-21 15:00:21.788036974 +0000
+++ binutils-2.27/ld/testsuite/ld-elf/shared.exp	2017-02-21 16:18:12.497641235 +0000
@@ -665,5 +665,5 @@ proc mix_pic_and_non_pic {xfails cflags
     }
 }
 
-mix_pic_and_non_pic [list "arm*-*-*" "aarch64*-*-*"] "" "" "pr19719"
+mix_pic_and_non_pic [list "arm*-*-*" "aarch64*-*-*" "s390*-*-*"] "" "" "pr19719"
 mix_pic_and_non_pic [] "-fPIE" "-pie" "pr19719pie"
