diff -rup binutils.orig/bfd/elf32-s390.c binutils-2.27/bfd/elf32-s390.c
--- binutils.orig/bfd/elf32-s390.c	2017-05-12 13:00:01.697650210 +0100
+++ binutils-2.27/bfd/elf32-s390.c	2017-05-12 13:01:05.223988744 +0100
@@ -2809,7 +2809,7 @@ elf_s390_relocate_section (bfd *output_b
 	      && s390_is_ifunc_symbol_p (h)
 	      && h->def_regular)
 	    {
-	      if (!bfd_link_pic (info) || !h->non_got_ref)
+	      if (!bfd_link_pic (info))
 		{
 		  /* For a non-shared object STT_GNU_IFUNC symbol must
 		     go through PLT.  */
diff -rup binutils.orig/bfd/elf64-s390.c binutils-2.27/bfd/elf64-s390.c
--- binutils.orig/bfd/elf64-s390.c	2017-05-12 13:00:01.699650190 +0100
+++ binutils-2.27/bfd/elf64-s390.c	2017-05-12 13:01:31.017720167 +0100
@@ -2768,10 +2768,11 @@ elf_s390_relocate_section (bfd *output_b
 	      && s390_is_ifunc_symbol_p (h)
 	      && h->def_regular)
 	    {
-	      if (!bfd_link_pic (info) || !h->non_got_ref)
+	      if (!bfd_link_pic (info))
 		{
-		  /* For a non-shared object STT_GNU_IFUNC symbol must
-		     go through PLT.  */
+		  /* For a non-shared object the symbol will not
+		     change.  Hence we can write the address of the
+		     target IPLT slot now.  */
 		  relocation = (htab->elf.iplt->output_section->vma
 				+ htab->elf.iplt->output_offset
 				+ h ->plt.offset);
diff -rup binutils.orig/bfd/elf-s390-common.c binutils-2.27/bfd/elf-s390-common.c
--- binutils.orig/bfd/elf-s390-common.c	2017-05-12 13:00:01.695650231 +0100
+++ binutils-2.27/bfd/elf-s390-common.c	2017-05-12 13:01:54.347477247 +0100
@@ -161,9 +161,7 @@ keep:
       h->type = STT_FUNC;
     }
 
-  /* We need dynamic relocation for STT_GNU_IFUNC symbol only when
-     there is a non-GOT reference in a shared object.  */
-  if (!bfd_link_pic (info) || !h->non_got_ref)
+  if (!bfd_link_pic (info))
     *head = NULL;
 
   /* Finally, allocate space.  */
